keps/2724-topology-aware-schedling/README.md

```diff
@@ -0,0 +1,672 @@
+)
+```
+
+### Validation
+
+The validations at the creation time:
+- the ResourceFlavor defining`TopologyAwareScheduling` needs to have at least
+  one node label
+
+We introduce the following validations (post-creation time, workload violating
+the rules is deactivated):
+- the value of `kueue.x-k8s.io/require-topology` is one of the labels specified
+  in the topology node labels
+- the value of `kueue.x-k8s.io/prefer-topology` is one of the labels specified
+  in the topology node labels
+
+### Internal APIs
+
+We extend the `Workload` structure to reflect the topology request at the
+Job level.
+
+```golang
+type WorkloadSpec struct {
+  ...
+  // TopologyRequest defines the topology requested for the corresponding Job.
+  TopologyRequest *TopologyRequest
+}
+
+type TopologyRequest struct {
+  // Policy defines the policy used for TAS. Possible values are:
+  // - Prefer set when `kueue.x-k8s.io/prefer-topology` annotation is set on the Job
+  // - Require set when `kueue.x-k8s.io/require-topology` annotation is set on the Job
+  Policy TopologyRequestPolicy
+
+  // Level indicated by the `kueue.x-k8s.io/prefer-topology` or `kueue.x-k8s.io/require-topology `
+  // annotation
+  Level string
+}
+```
+
+We extend the `PodSetAssignment` structure to keep track of the number of pods
+at each topology level to the specific subset of nodes.
+
+```golang
+type PodSetAssignment struct {
+  ...
+
+  // TopologyAssignment indicates the resources assigned per topology level
+  TopologyAssignment *TopologyAssignment
+}
+
+type TopologyAssignment struct {
+     // Slices contains the list of assignments split into slices
+     Slices []TopologyAssignmentSlice
+}
+
+type TopologyAssignmentSlice struct {
+  // NodeLabels constitutes the nodeSelector for a given slice of pods. It
+  // defines values for all labels configured in the Topology.Levels.
+  NodeLabels map[string]string
+
+  // Count indicates the number of pods in a given TopologyAssignmentSlice
+  Count int
+}
+```
+
+Kueue uses the `kueue.x-k8s.io/topology` scheduling gate to delay the
+`nodeSelector` assignment, because different pods in the same PodSet may have
+different values:
+
+```golang
+const (
+  // TopologySchedulingGate is used to delay topology assignment for pods
+  // once all the pods are created.
+  TopologySchedulingGate = "kueue.x-k8s.io/topology"
+
+  // WorkloadAnnotation indicates the name of the workload assigned.
+  WorkloadAnnotation = "kueue.x-k8s.io/workload"
+
+  // PodSetLabel indicates the name of the PodSet in the workload
+  PodSeLabel = "kueue.x-k8s.io/podset"
+)
+```
+
+### Computing the assignment
+
+The extended pod assignment is set on admitting the workload. In order to
+compute the assignment Kueue caches the information about the node allocatable
+capacity and the current usage by other workloads in this resource flavor.
+
+The information about the allocatable capacity is scraped from nodes based on
+the `status.allocatable` field.
```

Comment:
>Doesn't it violate the Kueue's responsibility?
> We used to delegate the node specific responsibility to ClusterAutoscaler and KubeScheduler.

Well. indeed this feature goes beyond just quota checking. 

We discussed with @mwielgus options to provide this feature via ClusterAutoscaler or KubeScheduler, but similarly the feature goes even more beyond responsibility of the mentioned components. I think Kueue is the best match we currently have.

> InterPodAffinity

I don't think you can achieve the feature here with `InterPodAffinity` without knowing what is the actual topology. There are many things which I don't think achivable otherwise, for example, the "prefer" semantics. We try "rack", but if not possible we fallback to "block" and try to optimize the number of used racks.

Reply:
Maybe add some notes about this decision in the Alternatives section.

I think the primary blocker is preemption. If the decision is given to cluster-autoscaler or kube-scheduler, then they won't have knowledge about what to preempt to make space, respecting priorities and/or fair sharing.

Reply:
Add a section in alternatives. @tenzen-y if this addresses your question I would resolve the comment.